### ISSUE TITLE
Show type info: use same format as in completion

### ISSUE THREAD

**Issue by ktonga** [ensime/ensime-vim#289]
Currently, `show type` command uses the already formatted type name from the server response. It is ok for simple types and `Funtion`s, but I find much more readable the format used in completion for methods.
We could reuse the same logic as in completion to give `ArrowTypeInfo` a special treatment.

I already have a TODO in the code for that. Let me know what you people think and I could open a PR for it.

**Comment by ches** [comment]
I'd have to see some "mock-up" and perhaps examples of types that show the value of it, personally. `:EnType` for instance is showing the inferred *result type* of an expression, which is usually going to be much simpler than the amount of information in call signatures. Even if you're inspecting something that's a function type, in that context I want it to be concise, I don't want my eyes to have to scan a mile to a second column on a wide monitor.

Take for example `(Int, String) => IndexedSeq[String]` from #288 – it doesn't get much better than that to me, it's literally a valid type that I can even copy and paste directly into `:kind` in the REPL.

Now `:EnInspectType`, that one I would be happy to have formatted differently, I'm just not sure how. It could even produce multiple lines of output requiring a keypress to dismiss IMO, to me this command is a choice to "dig deeper" than `:EnType`.

So that's just my perspective, not to discourage trying it, just need to visualize some cases to evaluate.

**Comment by ktonga** [comment]
You're totally right, I mean 'show type' action as a kinda 'quick doc' popup from the Idea world. But if it is just for inferred type I prefer the shortcut.

It would be nice to have a supporting split window (like vim preview window) to show this kind of information. Some times I want to quickly view the signature+scaladoc of a method. I don't want to switch to the browser for that.

I'll check what the server provides and sketch up a workflow for such use case. Let me rephrase this issue.

**Comment by ches** [comment]
> It would be nice to have a supporting split window (like vim preview window) to show this kind of information. Some times I want to quickly view the signature+scaladoc of a method. I don't want to switch to the browser for that.

That'd be great, this touches on two things I've had on my list too but haven't written up tickets for them yet:

- A lot of plugins like jedi-vim pop up a full function signature and possibly docstring in the preview window when completion is invoked, allowing you to keep seeing full function params after entering `(`. Not everyone likes that because it shifts your window layout, but I put `autocmd InsertLeave * pclose` in my vimrc and I'm quite happy with that setup. `set completeopt-='preview'` if you don't like it, should apply to all plugins that do this.

- Documentation lookup that doesn't require switching to a browser, to use for `K`. Probably open in a split rather than preview because you're likely to want to want to scroll it almost immediately, but this could be flexible. This might be clumsy since we'd probably just have to implement stripping HTML on-the-fly from the doc jars that ENSIME caches… but maybe someone has another idea.

**Comment by ktonga** [comment]
> A lot of plugins like jedi-vim pop up a full function signature and possibly docstring in the preview window when completion is invoked, allowing you to keep seeing full function params after entering (. Not everyone likes that because it shifts your window layout, but I put autocmd InsertLeave * pclose in my vimrc and I'm quite happy with that setup. set completeopt-='preview' if you don't like it, should apply to all plugins that do this.

I was aware of that, and I thought a lot about it while implementing #275, but no docs are available in the server response. So if there is an endpoint for docs we could do both requests and merge the responses to easily turn this feature on.
In the meantime, I had configured 'neosnipet' to help me jump through params after method completion. I even opened Shougo/neosnippet.vim#345 just in case to improve integration. I'd like to add some lines to the docs for providing some config examples to get a good autocompletion experience.

> Documentation lookup that doesn't require switching to a browser, to use for K. Probably open in a split rather than preview because you're likely to want to want to scroll it almost immediately, but this could be flexible. This might be clumsy since we'd probably just have to implement stripping HTML on-the-fly from the doc jars that ENSIME caches… but maybe someone has another idea.

I'll play around with the server to find what we can do. A feature like that would be so helpful.

**Comment by ktonga** [comment]
> Not everyone likes that because it shifts your window layout

Extra information for autocompletion would be much better if [something like this](https://github.com/neovim/neovim/issues/396#issuecomment-163760889) gets done:

**Comment by ktonga** [comment]
It seems that [some work](https://github.com/ensime/ensime-server/issues/888) is already planed on the server to support this use-case on the editors.
